### DISCUSSION TITLE
Reuse RID of deleted records

### DISCUSSION THREAD

**Discussion by lvca** [ArcadeData/arcadedb#1203] [Ideas]
ArcadeDB does not reuse the assigned RIDs. Never. Even when the record is deleted. This has the advantage to determine if a link is broken (the RID points to a non-existent record), useful in case of unidirectional edges.

In other cases reusing the RIDs allows reusing all the space in pages. In some use cases which are heavily based on deletes, this feature can save a lot of space and improve overall performance.

I'd like to let this configurable, so the user can decide per bucket/type, if the RIDs are reusable.

**Comment by vic0824** [comment]
Are the pages never compacted, like it happens for the indexes? If they aren't, then I agree it's useful to re-use the RIDs to optimize the space usage in pages.

**Comment by jtsmedley** [comment]
Maybe add to the `check database` command  suggestion of `RECLAIM` as a flag.

**Comment by ansarizafar** [comment]
My use case is heavily based on deletes so a solution is definitely required.

**Comment by ansarizafar** [comment]
Is there any new update?

**Reply by lvca** [reply]
We have a branch where we're able to reuse the RIDs, but this decreases the insertion speed when you have a lot of records in the buckets. We're testing a better approach where we save the deleted RIDs in a separate structure. It should be out on the January release.

**Reply by ansarizafar** [reply]
It means we will be able to reclaim the space of deleted records. If I start using  Arcadedb now, then would it be possible to upgrade my database to the new release?

**Reply by lvca** [reply]
Yes, it will recycle 100% of the deleted space. Upgrade will be compatible,  just update the jars

**Reply by ansarizafar** [reply]
Awesome.  It's a deal-breaker for my use case.

**Comment by ansarizafar** [comment]
I am still waiting for this feature.

**Reply by lvca** [reply]
We made good progress (it's in the `delete-reuse-space` branch), but it's still not stable enough to be merged into the main branch.
